### PR TITLE
allow downloading of recent versions of chrome

### PR DESCRIPTION
**Pull request by david-engelmann**
the hosting path for chrome has changed for recent versions

### CODE REVIEW

**Comment by david-engelmann** [comment]
need to patch File parsing of response, currently breaking because it's returning a file with .exe in the name

**Review by AjayIndana — APPROVED** [review]

**Comment by david-engelmann** [comment]
lets wait to merge, I'm still seeing some fails in CI/CD versions, I've almost resolved

**Comment by david-engelmann** [comment]
I've tested my little heart out, and it seems to be working locally and within my CI/CD pipelines. We should run the testing suite to confirm everything is good to go

**Comment by david-engelmann** [comment]
i adjust so that the failing 114.0.5735.198 test pass

**Review by mjblue — APPROVED** [review]
LGTM

**Review by vmred — APPROVED** [review]

**Comment by p112913** [comment]
hello, is there an ETA for the merge on this PR?

**Comment by david-engelmann** [comment]
@p112913 we should run the test suite again then merge it!

**Comment by p112913** [comment]
@david-engelmann I am assuming we need someone from maintainers to run the test suite and apply the merge?

**Comment by dimaklepikov** [comment]
@david-engelmann Thanks!

**Comment by morgan-young** [comment]
I've been watching this eagerly! Thanks for all your efforts @david-engelmann :+1:

**Comment by cguzmanceti** [comment]
Hello. I want to tell you that I found an additional issue for Mac with Apple Silicon. Inside your `Chrome` class, in the `drivers` folder, you return **'mac_arm64'** from `get_os_type()`, but on the Google Chrome Labs download page, it comes as **'mac-arm64'**, so it never matches and the URL is not sent correctly. Hopefully you can include this setting before you merge your branch. Thank you.

**`get_os_type()` method
Screenshot 2023-07-20 at 11 51 10

**Google Chrome Labs JSON**
Screenshot 2023-07-20 at 11 51 39

**Comment by david-engelmann** [comment]
@cguzmanceti I've added that fix to the PR. Would you be able to test on mac?

**Comment by cguzmanceti** [comment]
> @cguzmanceti I've added that fix to the PR. Would you be able to test on mac?

Thank you. It already let me download the driver, but now it shows the error **'cannot find Chrome binary'**. The strange thing is that when I change the `ChromeDriverManager().install()` declaration to the path where the driver is physically located, I can run **Selenium** without any problem.

**Comment by david-engelmann** [comment]
> > @cguzmanceti I've added that fix to the PR. Would you be able to test on mac?
> 
> Thank you. It already let me download the driver, but now it shows the error **'cannot find Chrome binary'**. The strange thing is that when I change the `ChromeDriverManager().install()` declaration to the path where the driver is physically located, I can run **Selenium** without any problem.

Do you have the chrome binaries for the version you tried to download?

**Comment by cguzmanceti** [comment]
> > > @cguzmanceti I've added that fix to the PR. Would you be able to test on mac?
> > 
> > 
> > Thank you. It already let me download the driver, but now it shows the error **'cannot find Chrome binary'**. The strange thing is that when I change the `ChromeDriverManager().install()` declaration to the path where the driver is physically located, I can run **Selenium** without any problem.
> 
> Do you have the chrome binaries for the version you tried to download?

In fact, use the path where WDM downloads the binaries

Screenshot 2023-07-20 at 13 09 53

**Comment by david-engelmann** [comment]
@cguzmanceti I have the chrome actually installed at "Current browser version is 115.0.5790.98 with binary path /Applications/Google Chrome.app/Contents/MacOS/Google Chrome" but the chrome install command from the webdriver_manager unzip the files to "/Users/YOUR_USERNAME/.wdm/drivers/chromedriver/mac_arm64/115.0.5790.98". I will continue to investigate

**Comment by cguzmanceti** [comment]
Thanks 😁

**Comment by david-engelmann** [comment]
@cguzmanceti when i run it from a venv on macos i get the following error
```
File "/Users/YOUR_USERNAME/projects/Python/webdriver_manager/webdriver_manager/chrome.py", line 40, in install
    os.chmod(driver_path, 0o755)
PermissionError: [Errno 1] Operation not permitted: '/Users/YOUR_USERNAME/.wdm/drivers/chromedriver/mac_arm64/115.0.5790.98/chromedriver'
```
the presence of the chromedriver file exist when not running from the venv. It seems to me that its a permission issue. I wonder if we need to handle it similarly to the class that handles unzipping in linux with permissions?

**Comment by phcamargo23** [comment]
How can I fix my application without waiting for the pull request to be accepted? 😥
I tried to upgrade the dependency directly via the pull request branch (pip install git+https://github.com/david-engelmann/webdriver_manager.git@de-upgrade-chrome-version), but it didn't work.

**Comment by cguzmanceti** [comment]
> @cguzmanceti when i run it from a venv on macos i get the following error
> 
> ```
> File "/Users/YOUR_USERNAME/projects/Python/webdriver_manager/webdriver_manager/chrome.py", line 40, in install
>     os.chmod(driver_path, 0o755)
> PermissionError: [Errno 1] Operation not permitted: '/Users/YOUR_USERNAME/.wdm/drivers/chromedriver/mac_arm64/115.0.5790.98/chromedriver'
> ```
> 
> the presence of the chromedriver file exist when not running from the venv. It seems to me that its a permission issue. I wonder if we need to handle it similarly to the class that handles unzipping in linux with permissions?

Hi. In my case, I didn't see that error. Possibly it is because the driver is installed from the same venv in which I have all my code.

**Comment by david-engelmann** [comment]
@cguzmanceti I get the following popup error when I try to load the chromedriver
Screen Shot 2023-07-20 at 6 04 01 PM

Seems like a pretty standard error in mac land. After resolving that issue by purging my chromedriver downloads, I returned to the libsystem_pthread.dylib error message.

Here is a similar issue in a different repo -  https://github.com/Significant-Gravitas/Auto-GPT/issues/1517. I'm not a big Mac user by choice so if anyone has more insight, please chim in

**Comment by p112913** [comment]
@david-engelmann this permission error is standard and is expected on Macs everytime a new chrome driver is downloaded. PS: I am a Mac user who has been testing selenium and web driver regularly

**Comment by david-engelmann** [comment]
> @david-engelmann this permission error is standard and is expected on Macs everytime a new chrome driver is downloaded. PS: I am a Mac user who has been testing selenium and web driver regularly

LGTM

**Comment by p112913** [comment]
@david-engelmann that is a different error and happens when web driver isn't able to find the chrome binary on a standard path e.g user has moved it to a diff path or if the machine does not have chrome installed at all.

**Comment by p112913** [comment]
it could also happen if the chrome binary does not have executable permission despite being present on the machine. I have run into that error a lot of times in the past and its one of these reasons.

**Comment by david-engelmann** [comment]
> @david-engelmann that is a different error and happens when web driver isn't able to find the chrome binary on a standard path

@p112913 I was testing with a version argument, I thought it was a permission issue but I saw 644 when I investigated the chromedriver file path permission code. Originally, I thought it might be that because when I ran `ls -l` it was in a red font

**Comment by david-engelmann** [comment]
@p112913 @cguzmanceti I dug a bit deeper 

when we unzip the chromedriver, we have the following perms
```
save_file_to_cache path: /Users/david/.wdm/drivers/chromedriver/mac_arm64/115.0.5790.98
name: chromedriver
file perm mask: 644
```
However, when I run the same test with the results of ChromeDriverManager().install(), I got the following perms
```
/Users/david/.wdm/drivers/chromedriver/mac_arm64/115.0.5790.98/chromedriver
755
```
Install runs the following command
```
os.chmod(driver_path, 0o755)
```
We can avoid making that chmod but I'm worried about it's effect on windows and linux, we can inspect the driver_path for mac_arm64 and if it's not in the path then make the chmod.

**Comment by zaelcovsky** [comment]
@david-engelmann Sorry for lame question - what is the relation between Chromedriver and Chrome browser in os?  First, we need Chrome browser being inastalled  for run Сhromedriver, right? Maybe there are some additional requirements? Is it any relation between version of Chromedriver and version of  Chrome browser in our os, like they need to be identical?

**Comment by Derpitron** [comment]
Will I have to change or fix my code, when this PR is merged?

**Comment by mathiasbynens** [comment]
FWIW, here was the official announcement for the ChromeDriver changes: https://groups.google.com/g/chromedriver-users/c/clpipqvOGjE/m/5NxzS_SRAgAJ

**Comment by skorokithakis** [comment]
I haven't been following the entire thread, but is it possible to use the last webdriver on network issues? That way it won't fail if something similar happens again.

**Comment by SudhanshuBlaze** [comment]
When can we expect this PR to get merged? Any approximation 😄

**Comment by david-engelmann** [comment]
> I haven't been following the entire thread, but is it possible to use the last webdriver on network issues? That way it won't fail if something similar happens again.

>Will I have to change or fix my code, when this PR is merged? Or, will this fix essentially be invisible to dependents?

It should have the same desired behavior, just have the ability to work with new versions of chrome

File: webdriver_manager/drivers/chrome.py
Comment (p112913) [inline webdriver_manager/drivers/chrome.py:48]: @david-engelmann this version parse should be >= "115" because that is where the new download url comes into play.

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py:48]: @p112913 check out the past action results and this comment https://github.com/SergeyPirogov/webdriver_manager/pull/537#issuecomment-1643671232. 114.0.5735.198 isn't present in the old url

Comment (p112913) [reply webdriver_manager/drivers/chrome.py:48]: @david-engelmann using the new url I get a 404 for this version:
https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.198/mac-arm64/chromedriver-mac-arm64.zip

Also the last known version I see on release 114 is 114.0.5735.90 and even for that the new url doesn't work. You can try this in your browser and you would get back a 404:

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py:48]: gotcha so it looks like the testing pipelines has a bad test for 114.0.5735.1988. I'll update the version check

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py]: Failed test in question
```
FAILED tests/test_brave_driver.py::test_brave_manager_with_selenium - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/115/chromedriver_linux64.zip
FAILED tests/test_chrome_driver.py::test_chrome_manager_with_cache - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_linux64.zip
FAILED tests/test_chrome_driver.py::test_chrome_manager_with_selenium - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_linux64.zip
FAILED tests/test_chrome_driver.py::test_driver_with_ssl_verify_disabled_can_be_downloaded - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_linux64.zip
FAILED tests/test_chrome_driver.py::test_chrome_manager_cached_driver_with_selenium - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_linux64.zip
FAILED tests/test_chrome_driver.py::test_can_get_chrome_for_win[win32] - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_win32.zip
FAILED tests/test_chrome_driver.py::test_can_get_chrome_for_win[win64] - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_win32.zip
FAILED tests/test_chrome_driver.py::test_can_get_chrome_for_win[mac64] - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_mac_arm64.zip
FAILED tests/test_chrome_driver.py::test_can_get_chrome_for_win[mac64_m1] - ValueError: There is no such driver by url https://chromedriver.storage.googleapis.com/114.0.5735.198/chromedriver_mac_arm64.zip
```

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py]: I see the last known 114.0.5735.90 available for download. The filename was different then the ones you tried. I'm still ok with starting with version 115 for the new endpoint. One thing to consider, people used to pass versions="114" etc. I'm not sure that format works with the new endpoint

here is the json block 
```
{
      "version": "114.0.5735.90",
      "revision": "1135570",
      "downloads": {
        "chrome": [
          {
            "platform": "linux64",
            "url": "https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.90/linux64/chrome-linux64.zip"
          },
          {
            "platform": "mac-arm64",
            "url": "https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.90/mac-arm64/chrome-mac-arm64.zip"
          },
          {
            "platform": "mac-x64",
            "url": "https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.90/mac-x64/chrome-mac-x64.zip"
          },
          {
            "platform": "win32",
            "url": "https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.90/win32/chrome-win32.zip"
          },
          {
            "platform": "win64",
            "url": "https://edgedl.me.gvt1.com/edgedl/chrome/chrome-for-testing/114.0.5735.90/win64/chrome-win64.zip"
          }
        ]
      }
    },
```

Comment (p112913) [reply webdriver_manager/drivers/chrome.py]: oh, so I was missing the "chromedriver" at the end in my url. I think since versions below 115 are still accessible on the old url so let's use the old url for all versions below 115 and then starting >= 115 we can use the new url.

Comment (p112913) [reply webdriver_manager/drivers/chrome.py]: for passing of version I don't know how it will supported by the new url format. I am not sure of what the logic is for versions 114 and below but I think it we should retain that feature perhaps with versions >=115 ? I can see some more work probably needs to be put in to have that feature supported with the new url

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py]: we can use https://googlechromelabs.github.io/chrome-for-testing/last-known-good-versions.json similarly to the LATEST_RELEASE endpoint of old

Comment (mathiasbynens) [reply webdriver_manager/drivers/chrome.py]: Hi, ChromeDriver lead here. The new locations only apply to 115 and later. For 114 and below the old URLs still apply.

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py]: @mathiasbynens Perfect, I've just added support for the simple version ie. 115 and the build version ie. 115.0.5735

**Comment by tonycapone** [comment]
Hi @david-engelmann. Tried out your repo and branch to get around the url issue,
`pip install git+https://github.com/david-engelmann/webdriver_manager.git@de-upgrade-chrome-version`


But I get this error. Any ideas?

```
  File "/usr/local/lib/python3.10/site-packages/webdriver_manager/chrome.py", line 39, in install
    driver_path = self._get_driver_path(self.driver)
  File "/usr/local/lib/python3.10/site-packages/webdriver_manager/core/manager.py", line 30, in _get_driver_path
    file = self._download_manager.download_file(driver.get_driver_download_url())
  File "/usr/local/lib/python3.10/site-packages/webdriver_manager/drivers/chrome.py", line 63, in get_driver_download_url
    release_version = self.get_latest_patch_version_for_build_version(build_version=driver_version_to_download)
  File "/usr/local/lib/python3.10/site-packages/webdriver_manager/drivers/chrome.py", line 117, in get_latest_patch_version_for_build_version
    return data["builds"][build_version]["version"]
KeyError: '115.0.5790.102'
```

**Comment by david-engelmann** [comment]
> KeyError: '115.0.5790.102'

I'm not sure that is a supported version

**Comment by david-engelmann** [comment]
@tonycapone I believe that should work now

**Comment by SergeyPirogov** [comment]
Thanks @david-engelmann, could you add a specific test for chrome v115?

**Comment by SergeyPirogov** [comment]
I've merged fixes for chrome. Please test and provide a feedback

**Comment by david-engelmann** [comment]
> Thanks @david-engelmann, could you add a specific test for chrome v115?

@SergeyPirogov It's unclear how I define test cases for certain chrome versions. the test failing on 114 is because that version (114.0.5735.198) doesn't exist. I was looking at your merge conflicts and I couldn't follow the logic for most of them. In particular, the removal of support for the short form version 115 and the build version 115.0.5735 etc. I can remove that logic but it will require users to use the patched version of their browser which changes frequently.

**Comment by SergeyPirogov** [comment]
@david-engelmann I've fixed this in github actions by installing latest Google Chrome for every os

**Comment by david-engelmann** [comment]
@SergeyPirogov Perfect! I adjusted the parameters and os_type logic. The only conflict I left in was the short form support for version xxx and version xxx.x.xxxx.

**Comment by SergeyPirogov** [comment]
@david-engelmann please point to exect lines of code, because its hard to get what do you mean by short version

File: webdriver_manager/drivers/chrome.py
Comment (david-engelmann) [inline webdriver_manager/drivers/chrome.py:112]: @SergeyPirogov Here is the function for latest release by short version ie. 115

File: webdriver_manager/drivers/chrome.py
Comment (david-engelmann) [inline webdriver_manager/drivers/chrome.py:122]: @SergeyPirogov Here is the function for latest release by build version ie. 115.0.5735

Comment (SergeyPirogov) [reply webdriver_manager/drivers/chrome.py:122]: So, why do we need it?

Comment (david-engelmann) [reply webdriver_manager/drivers/chrome.py:122]: @SergeyPirogov The old LATEST_RELEASE url supported those style lookups. ie. https://chromedriver.storage.googleapis.com/LATEST_RELEASE_113 and https://chromedriver.storage.googleapis.com/LATEST_RELEASE_114.0.5735. [It was suggested we retain that for versions over 115](https://github.com/SergeyPirogov/webdriver_manager/pull/537#discussion_r1270761387). Without that, users will have to pass in the full browser version ie. version="114.0.5735.90", which can change frequently. Either way, I'm fine including for backwards compatibility, I'm fine with removing for simplicity and add an exception that explains as suggested by you

Comment (SergeyPirogov) [reply webdriver_manager/drivers/chrome.py:118]: I'd postpone this, because what I see know is that browser version is equal to the one mentioned in their driver website. Moreover I think that fix should be applied in method **get_url_for_version_and_platform**

**Comment by david-engelmann** [comment]
Code merged in https://github.com/SergeyPirogov/webdriver_manager/pull/539

**Comment by david-engelmann** [comment]
going to close out and submit a new pr for the [short version support feature discussed here](https://github.com/SergeyPirogov/webdriver_manager/pull/537#discussion_r1271318429)

**Comment by p112913** [comment]
Thanks @david-engelmann for all your hard work on this PR!

**Comment by grapefruiteater** [comment]
Sorry, Chrome changed driver url path.
https://googlechromelabs.github.io/chrome-for-testing/

please re-revise content.

**Comment by grapefruiteater** [comment]
This is my error. Driver path 115.0.5790.110 is not valid from above URL site.

Traceback (most recent call last):
  File "C:\Users\11235\Documents\Workfolder\Amazon_GetOrders\\Amazon_GetOrders.py", line 73, in <module>
    driver_path = ChromeDriverManager().install()
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\11235\AppData\Local\Programs\Python\Python311\Lib\site-packages\webdriver_manager\chrome.py", line 39, in install
    driver_path = self._get_driver_binary_path(self.driver)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\11235\AppData\Local\Programs\Python\Python311\Lib\site-packages\webdriver_manager\core\manager.py", line 33, in _get_driver_binary_path
    file = self._download_manager.download_file(driver.get_driver_download_url())
                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\11235\AppData\Local\Programs\Python\Python311\Lib\site-packages\webdriver_manager\drivers\chrome.py", line 59, in get_driver_download_url
    modern_version_url = self.get_url_for_version_and_platform(driver_version_to_download, os_type)
                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\11235\AppData\Local\Programs\Python\Python311\Lib\site-packages\webdriver_manager\drivers\chrome.py", line 94, in get_url_for_version_and_platform
    raise Exception(f"No such driver version {browser_version} for {platform}")
Exception: No such driver version 115.0.5790.110 for win32

**Comment by david-engelmann** [comment]
@grapefruiteater It appears that the version of Chrome you have is not supported by chrome-for-testing

**Comment by Derpitron** [comment]
> @grapefruiteater It appears that the version of Chrome you have is not supported by chrome-for-testing

I am experiencing this issue as well. How can I fix it?

**Comment by david-engelmann** [comment]
@Derpitron I would install a version of chrome that matches a version supported by chrome-for-testing. [This link](https://googlechromelabs.github.io/chrome-for-testing/known-good-versions-with-downloads.json) shows everything they support. The main issue is the patch version -> 115.0.5790.**_**110**_**. 102 is the last version supported by google-for-testing

**Pull request closed without merging**
